From a7bc0ce187447843bbdd46f72dadc645fb72b733 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Fri, 18 Oct 2024 10:15:25 +0200
Subject: [PATCH 1332/1695] clk: stm32mp25: check RIF access for HSI/HER/MSI
 Kernel clocks

Check RIF access on clock enable/disable for HSI/HSE/MSI Kernel clocks.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 8621d232e7b0a69748b6f318cc0831129a9824b0 clk: stm32mp25: check RIF access for HSI/HER/MSI Kernel clocks]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I2fbc00f60a8c74f7551d2427882002b6cd61ea6a
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/412291
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/drivers/clk/clk-stm32mp25.c | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/core/drivers/clk/clk-stm32mp25.c b/core/drivers/clk/clk-stm32mp25.c
index b689bf887..57056181a 100644
--- a/core/drivers/clk/clk-stm32mp25.c
+++ b/core/drivers/clk/clk-stm32mp25.c
@@ -2497,6 +2497,26 @@ static const struct clk_ops clk_stm32_oscillator_msi_ops = {
 	.restore_context = clk_stm32_osc_msi_pm_restore,
 };
 
+static TEE_Result clk_stm32_osc_ker_enable(struct clk *clk)
+{
+	if (stm32_rcc_has_access_by_id(RCC_RIF_OSCILLATORS))
+		return clk_stm32_gate_enable(clk);
+
+	return TEE_SUCCESS;
+}
+
+static void clk_stm32_osc_ker_disable(struct clk *clk)
+{
+	if (stm32_rcc_has_access_by_id(RCC_RIF_OSCILLATORS))
+		clk_stm32_gate_disable(clk);
+}
+
+static const struct clk_ops clk_stm32_osc_ker_ops = {
+	.enable		= clk_stm32_osc_ker_enable,
+	.disable	= clk_stm32_osc_ker_disable,
+	.is_enabled	= clk_stm32_gate_is_enabled,
+};
+
 static TEE_Result clk_stm32_hse_div_set_rate(struct clk *clk,
 					     unsigned long rate,
 					     unsigned long parent_rate)
@@ -3417,7 +3437,7 @@ static const struct clk_ops ck_timer_ops = {
 
 #define STM32_OSC_KER(_name, _flags, _gate_id)\
 	struct clk _name = {\
-		.ops = &clk_stm32_gate_ops,\
+		.ops = &clk_stm32_osc_ker_ops,\
 		.priv = &(struct clk_stm32_gate_cfg){\
 			.gate_id = (_gate_id),\
 		},\
-- 
2.47.2

